### PR TITLE
Feature/object avoidance

### PR DESCRIPTION
**Pull request by saina-ramyar**
# PR Details
## Description
Update the in-lane cruising plugin to include object avoidance logic.

## Related Issue
#834 

## How Has This Been Tested?
The changes are unit tested and integration tested on the Black Pacifica

## Types of changes

- [x] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp [hunk 1]
```diff
@@ -94,7 +94,16 @@ bool InLaneCruisingPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest& r
   trajectory.trajectory_points = compose_trajectory_from_centerline(downsampled_points, req.vehicle_state); // Compute the trajectory
   trajectory.initial_longitudinal_velocity = std::max(req.vehicle_state.longitudinal_vel, config_.minimum_speed);
 
-  resp.trajectory_plan = trajectory;
+  if (config_.enable_object_avoidance)
+  {
+    object_avoidance::ObjectAvoidance obj;
```
Comment (kjrush) [inline inlanecruising_plugin/src/inlanecruising_plugin.cpp:99]: Is there a reason to create and recreate this on a timestep-by timestep basis? I think we could just initialize it or set it to `nullptr` in the constructor (or wherever this param is loaded) and then check here to see if it's non-null.

Comment (kjrush) [reply inlanecruising_plugin/src/inlanecruising_plugin.cpp:99]: Either that or make it a free function and avoid having to create an object at all, if there's no state here that matters.

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
+
+#include <inlanecruising_plugin/object_avoidance.h>
+
+namespace inlanecruising_plugin
+{
+namespace object_avoidance
+{
+    ObjectAvoidance::ObjectAvoidance(){}
+
+    cav_msgs::TrajectoryPlan ObjectAvoidance::update_traj_for_object(cav_msgs::TrajectoryPlan& original_tp, const carma_wm::WorldModelConstPtr& wm_, double current_speed_)
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:10]: Can the first param be made const?

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        current_velocity.linear.x = current_speed_;
+
+        // // TODO get roadway object
+        std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+        cav_msgs::RoadwayObstacleList rwol2;
+        rwol2.roadway_obstacles = rwol;
+
+        std::cout << "host_vehicle_size" << host_vehicle_size.x << std::endl;
+
+        std::vector<cav_msgs::RoadwayObstacle> rwol_collision = carma_wm::collision_detection::WorldCollisionDetection(rwol2, original_tp, host_vehicle_size, current_velocity, 10.0);// 10 replace to target_time
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:25]: Does it make sense to set target time to be the duration of the trajectory plan?

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        if(rwol_collision.size() > 0) {
+
+            std::cout << "Collision!" << std::endl;
+
+            // use trajectory utiles to update trajectory plan points
+            
+            // lead vehicle trjactory
+            double dist_x = rwol_collision[0].object.pose.pose.position.x - original_tp.trajectory_points[0].x;
+            double dist_y = rwol_collision[0].object.pose.pose.position.y - original_tp.trajectory_points[0].y;
+            double x_lead = sqrt(dist_x*dist_x + dist_y*dist_y);
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:39]: What frame are dist_x and dist_y in? I'm not sure I understand either what value this represents or how it's calculated. I'm thinking this is the distance from the ego-vehicle to the collision point, but there's no reference to the ego vehicle's position? Can you elaborate a bit more what's going on here?

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+            // roadway object position
+            double gap_time = (x_lead - x_gap)/current_speed_;
+
+            double collision_time = 0; //\TODO comming from carma_wm collision detection in future (not used now)
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:44]: If we need to update this in the future, let's get a Github and JIRA issue written up to follow up on these TODOs.

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                tp = t_ph;
+            }
+            else if(t_ph < tpmin){
+                tp = tpmin;
+            }
+            else {
+                tp = t_ref;
+            }
+
+            std::cout << "tp:  "<< tp << std::endl;
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:76]: If these are worth preserving, let's make them roslog debug statements

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            double initial_accel = 0;
+            double goal_accel = 0;
+
+            // TODO: Add special case for when the object is not moving
+
+            double delta_v_max = abs(rwol_collision[0].object.velocity.twist.linear.x - max_trajectory_speed(original_tp.trajectory_points));
+            
+            double t_ref = (original_tp.trajectory_points[original_tp.trajectory_points.size() - 1].target_time.toSec() - original_tp.trajectory_points[0].target_time.toSec());
+
+            double t_ph = 4 * delta_v_max / maximum_deceleration_value;
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:62]: These short variable names make this a bit hard to parse at first pass, I'd recommend either making them a bit more descriptive or adding some more comments to describe what's going on here.

Comment (kjrush) [reply inlanecruising_plugin/src/object_avoidance.cpp:62]: Also, it's not clear what that `4` literal is for, is it just an adjustment factor (to allow 4 times the minimum time necessary to decelerate?) If so, might be worth making it a param. Otherwise might not hurt to make it a named constant for clarity.

File: inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+#include <trajectory_utils/quintic_coefficient_calculator.h>
+
+
+
+namespace inlanecruising_plugin
+{
+namespace object_avoidance
+{
+
+    class ObjectAvoidance
```
Comment (kjrush) [inline inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h:43]: Member variables here should have the `_` marking them as such. Might be worth considering which can be made const as well, or which might be good as params.

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+            double tp = 0;
+
+            if(t_ph > tpmin && t_ref < t_ph){
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:66]: Is this supposed to be `... && t_ph < t_ref)`? Looks like you're trying to clamp tp to be between `tpmin` (lower bound) and `t_ref` (upper bound), but I think this inequality is inverted for that.

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+            for(size_t i = 1; i < original_tp.trajectory_points.size(); i++ )
+            {            
+                double traj_target_time = i * tp / original_tp.trajectory_points.size();
+                std::cout << "traj_target_time: "<< traj_target_time << std::endl;
+                double dt_dist = polynomial_calc(values, traj_target_time);
+                double dv = polynomial_calc_d(values, traj_target_time);
+
+                std::cout << "new dist: "<< dt_dist << std::endl;
+                std::cout << "new velocity: "<< dv << std::endl; //TODO Consider case when resulting velocity is 0 or <0
```
Comment (kjrush) [inline inlanecruising_plugin/src/object_avoidance.cpp:102]: For these kinds of edge cases that we're leaving unconsidered for now, let's add checks and warn-level logs so that it's obvious if we're looking for it.

**Review by kjrush — CHANGES_REQUESTED** [review]

File: inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        public:
+        ObjectAvoidance();
+
+
+        geometry_msgs::Vector3 host_vehicle_size;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h:49]: I made it public so can set the parameter for the unit test. It will be config param in yield.

File: inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        double polynomial_calc_d(std::vector<double> coeff, double x);
+
+        // finds maximum speed in a set of trajectory points
+        double max_trajectory_speed(std::vector<cav_msgs::TrajectoryPlanPoint> trajectory_points);
+
+        // calculates distance between trajectory points in a plan
+        std::vector<double> get_relative_downtracks(cav_msgs::TrajectoryPlan& trajectory_plan);
+
+        // parameter for activating object avoidance logic
+        bool enable_avoidance = true;
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h:66]: removed. Do not need it inside object_avoidance

File: inlanecruising_plugin/src/object_avoidance.cpp [hunk 13]
```diff
@@ -0,0 +1,189 @@
+
```
[the remark this answers is not in the document]
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:1]: added

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp (hunk 1 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/inlanecruising_plugin.cpp:99]: addressed, but did not add it to the constructor because that would fail all the ilc unit tests.

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 2 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:10]: added

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 3 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:25]: I think the trajectory time (6 s) is not sufficient and we would want to detect a collision and start slowing down earlier. But changed that into a parameter that we can change.

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 5 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:44]: created CAR 4288

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 6 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:76]: added

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 7 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:62]: addressed

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 9 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:66]: updated

File: inlanecruising_plugin/src/object_avoidance.cpp (hunk 10 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/src/object_avoidance.cpp:102]: added the warn. Also, these are being addressed in Yied

File: inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h (hunk 8 above)
Comment (saina-ramyar) [reply inlanecruising_plugin/include/inlanecruising_plugin/object_avoidance.h:43]: added _ and const. These are params in Yield, but not here, since it would include making changes to the constructor and also more cleanup after removing.

**Review by kjrush — APPROVED** [review]
